### ISSUE TITLE
C support

### ISSUE THREAD

**Issue by BartMassey** [dtolnay/cxx#4]
It would be great to provide a feature or something for this crate so that it could generate safe C bindings rather than C++. I would be willing to at least take a look at this; advice on how to proceed would be welcome.

**Comment by dtolnay** [comment]
I agree. For the next couple weeks at least, I think I am going to have to focus mostly on C++, but I absolutely intend to come back to C support.

For now the most helpful would be if someone were to fork and get something simple working with C. That would give an understanding of the magnitude of differences needed and a head start on some of the design work.

Some of the known work I can call out up front:

- Currently for C++ we translate Rust references like `&T` into C++ references `const T&`. For C presumably we would want those same signatures to be emitted using pointers.

- What to do about owned indirect values. I anticipate our opaque types concept will continue to play an important role in C as in C++, and currently the way we pass ownership of those across the FFI is restricted to Box and unique_ptr. It's possible we may be able to lift this restriction and pass inline by value, or we may need a C-idiomatic binding that matches Box's abi.

I don't expect these to be figured out here in the thread but they would be some of the first things to focus on in a fork.

**Comment by BartMassey** [comment]
Apologies. It is becoming clear to me that I wont have time to work on this anytime soon. I *am* excited to see it done, and hope someone will pick it up.
